File: DevEnvironmentCLion/Base/core/MessageWriter.cpp
#include "PreCompiled.h"
#include "MessageWriter.h"
#include "WindowsConsoleColorHandler.h"
#include "ConsoleColorSetter.h"

using namespace base;

/**
 * Creates a new message writer.
 */
IMessageWriterSharedPtr MessageWriter::make()
{
    return std::make_shared<MessageWriter>(
        std::cout,
        std::cout,
        std::cout,
        std::cout,
        createConsoleColorHandler());
}

/**
 * Creates a new message writer with no console color handler.
 */
IMessageWriterSharedPtr MessageWriter::makeLight()
{
    return std::make_shared<MessageWriter>(
        std::cout,
        std::cout,
        std::cout,
        std::cout,
        nullptr);
}

/**
 * The MessageWriter constructor.
 */
MessageWriter::MessageWriter(
    std::ostream& informationalStream,
    std::ostream& warningStream,
    std::ostream& errorStream,
    std::ostream& debugStream,
    base::IConsoleColorHandlerSharedPtr consoleColorHandler) :
    m_informationalStream(informationalStream),
    m_warningStream(warningStream),
    m_errorStream(errorStream),
    m_debugStream(debugStream),
    m_consoleColorHandler(consoleColorHandler)
{
}

/**
 * The MessageWriter destructor.
 */
MessageWriter::~MessageWriter()
{
}

/**
 * Writes an informational message.
 */
void MessageWriter::writeInformationalMessage(const std::string& message)
{
    base::ConsoleColorSetter consoleColorSetter(
        m_consoleColorHandler,
        base::StandardFileDescriptor::StandardOutput,
        ColorType::Green);

    writeMessage(message, m_informationalStream);
}

/**
 * Writes a warning message.
 */
void MessageWriter::writeWarningMessage(const std::string& message)
{
    base::ConsoleColorSetter consoleColorSetter(
        m_consoleColorHandler,
        base::StandardFileDescriptor::StandardError,
        ColorType::Yellow);

    writeMessage(message, m_warningStream);
}

/**
 * Writes an error message.
 */
void MessageWriter::writeErrorMessage(const std::string& message)
{
    base::ConsoleColorSetter consoleColorSetter(
        m_consoleColorHandler,
        base::StandardFileDescriptor::StandardError,
        ColorType::Green);

    writeMessage(message, m_errorStream);
}

/**
 * Writes a debug message.
 */
void MessageWriter::writeDebugMessage(const std::string& message)
{
    base::ConsoleColorSetter consoleColorSetter(
        m_consoleColorHandler,
        base::StandardFileDescriptor::StandardOutput,
        ColorType::Green);

    writeMessage(message, m_debugStream);
}

/**
 * Writes a message to an output stream.
 */
void MessageWriter::writeMessage(
    const std::string& message,
    std::ostream& ostream)
{
    ostream << message << std::endl;
}

/**
 * Creates a console color handler.
 */
base::IConsoleColorHandlerSharedPtr MessageWriter::createConsoleColorHandler()
{
    #if defined(WIN32)
        return std::make_shared<base::WindowsConsoleColorHandler>();
    # else
        return nullptr;
    #endif
}

File: DevEnvironment/Base/src/main/java/base/core/MessageWriter.java
package base.core;

import base.interfaces.IDestructorHandler;
import base.interfaces.IMessageWriter;
import java.io.Closeable;
import java.io.PrintWriter;

/**
 * The MessageWriter class implements a writer of messages.
 */
public final class MessageWriter implements Closeable, IMessageWriter {
    private final PrintWriter informationalWriter;
    private final PrintWriter warningWriter;
    private final PrintWriter errorWriter;
    private final IDestructorHandler destructorHandler;

    /**
     * The MessageWriter constructor.
     */
    public MessageWriter() {
        try (DestructorHandler destructorHandler = new DestructorHandler()) {
            PrintWriter informationalWriter = new PrintWriter(System.out);
            destructorHandler.register(informationalWriter);
            PrintWriter warningWriter = new PrintWriter(System.err);
            destructorHandler.register(warningWriter);
            PrintWriter errorWriter = new PrintWriter(System.err);
            destructorHandler.register(errorWriter);

            this.destructorHandler = destructorHandler.move();

            this.informationalWriter = informationalWriter;
            this.warningWriter = warningWriter;
            this.errorWriter = errorWriter;
        }
    }

    /**
     * Closes this resource.
     */
    @Override
    public void close() {
        this.destructorHandler.close();
    }

    /**
     * Writes an informational message.
     */
    @Override
    public void writeInformationalMessage(String message) {
        this.writeMessage(
            this.informationalWriter,
            ColorType.GREEN + message);
    }

    /**
     * Writes a warning message.
     */
    @Override
    public void writeWarningMessage(String message) {
        this.writeMessage(
            this.warningWriter,
            ColorType.YELLOW + message);
    }

    /**
     * Writes an error message.
     */
    @Override
    public void writeErrorMessage(String message) {
        this.writeMessage(
            this.errorWriter,
            ColorType.RED + message);
    }

    /**
     * Gets an error writer.
     */
    @Override
    public PrintWriter getErrorWriter() {
        return this.errorWriter;
    }

    /**
     * Gets a warning writer.
     */
    @Override
    public PrintWriter getWarningWriter() {
        return this.warningWriter;
    }

    /**
     * Gets an informational writer.
     */
    @Override
    public PrintWriter getInformationalWriter() {
        return this.informationalWriter;
    }

    /**
     * Writes a message to an output stream.
     */
    private void writeMessage(PrintWriter writer, String message) {
        writer.println(message);
        writer.flush();
    }
}
